Two System.Data (and verify) tests fail when BCL is compiled with mcs (throw expression followed by ret)

Seems to be invalid IL being generated. My suspicion is that mcs might be generating invalid IL, rather than a PPC JIT issue. Unfortunately, I can't validate on my amd64 system because the test suite consumes all memory until killed. (WTF) Could someone verify with an x86/ARM system with the BCL compiled with mcs and run the System.Data test suite?

This is reproducible on Linux/ppc64le and AIX/ppc64be. (Retitle if it's an mcs issue reproducible elsewhere.)

```
Tests run: 2405, Passed: 2052, Errors: 2, Failures: 0, Inconclusive: 0
  Not run: 351, Invalid: 0, Ignored: 351, Skipped: 0
Elapsed time: 00:00:08.3300000

Errors and Failures:

1) BeginTransaction_Connection_Closed (MonoTests.System.Data.Odbc.OdbcConnectionTest.BeginTransaction_Connection_Closed)
   System.InvalidProgramException : Invalid IL code in System.Data.ProviderBase.DbConnectionClosed:BeginTransaction (System.Data.IsolationLevel): IL_0006: ret       


  at System.Data.Odbc.OdbcConnection.BeginTransaction (System.Data.IsolationLevel isolevel) [0x00000] in /home/builder/jenkins/workspace/test-mono-mainline-community-chroot@3/label/debian-9-ppc64el/external/corefx/src/System.Data.Odbc/src/System/Data/Odbc/OdbcConnection.cs:288 
  at System.Data.Odbc.OdbcConnection.BeginTransaction () [0x00000] in /home/builder/jenkins/workspace/test-mono-mainline-community-chroot@3/label/debian-9-ppc64el/external/corefx/src/System.Data.Odbc/src/System/Data/Odbc/OdbcConnection.cs:283 
  at (wrapper remoting-invoke-with-check) System.Data.Odbc.OdbcConnection.BeginTransaction()
  at MonoTests.System.Data.Odbc.OdbcConnectionTest.BeginTransaction_Connection_Closed () [0x00006] in /home/builder/jenkins/workspace/test-mono-mainline-community-chroot@3/label/debian-9-ppc64el/mcs/class/System.Data/Test/System.Data.Odbc/OdbcConnectionTest.cs:89 
  at (wrapper managed-to-native) System.Reflection.MonoMethod.InternalInvoke(System.Reflection.MonoMethod,object,object[],System.Exception&)
  at System.Reflection.MonoMethod.Invoke (System.Object obj, System.Reflection.BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) [0x00044] in /home/builder/jenkins/workspace/test-mono-mainline-community-chroot@3/label/debian-9-ppc64el/mcs/class/corlib/System.Reflection/MonoMethod.cs:305 


2) ServerVersion_Connection_Closed (MonoTests.System.Data.Odbc.OdbcConnectionTest.ServerVersion_Connection_Closed)
   System.InvalidProgramException : Invalid IL code in System.Data.ProviderBase.DbConnectionClosed:get_ServerVersion (): IL_0006: ret       


  at System.Data.Odbc.OdbcConnection.get_ServerVersion () [0x00000] in /home/builder/jenkins/workspace/test-mono-mainline-community-chroot@3/label/debian-9-ppc64el/external/corefx/src/System.Data.Odbc/src/System/Data/Odbc/OdbcConnection.cs:130 
  at MonoTests.System.Data.Odbc.OdbcConnectionTest.ServerVersion_Connection_Closed () [0x00006] in /home/builder/jenkins/workspace/test-mono-mainline-community-chroot@3/label/debian-9-ppc64el/mcs/class/System.Data/Test/System.Data.Odbc/OdbcConnectionTest.cs:215 
  at (wrapper managed-to-native) System.Reflection.MonoMethod.InternalInvoke(System.Reflection.MonoMethod,object,object[],System.Exception&)
  at System.Reflection.MonoMethod.Invoke (System.Object obj, System.Reflection.BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) [0x00044] in /home/builder/jenkins/workspace/test-mono-mainline-community-chroot@3/label/debian-9-ppc64el/mcs/class/corlib/System.Reflection/MonoMethod.cs:305
```

Just clean building I get:

```
Unhandled Exception:
Mono.Cecil.Cil.SymbolsNotMatchingException: Symbols were found but are not matching the assembly
  at Mono.Cecil.ModuleDefinition.ReadSymbols (Mono.Cecil.Cil.ISymbolReader reader, System.Boolean throwIfSymbolsAreNotMaching) [0x00045] in <64887e5183394bbb92d4e93b00abc715>:0 
  at Mono.Cecil.ModuleReader.ReadSymbols (Mono.Cecil.ModuleDefinition module, Mono.Cecil.ReaderParameters parameters) [0x00064] in <64887e5183394bbb92d4e93b00abc715>:0 
  at Mono.Cecil.ModuleReader.CreateModule (Mono.Cecil.PE.Image image, Mono.Cecil.ReaderParameters parameters) [0x0008e] in <64887e5183394bbb92d4e93b00abc715>:0 
  at Mono.Cecil.ModuleDefinition.ReadModule (Mono.Disposable`1[T] stream, System.String fileName, Mono.Cecil.ReaderParameters parameters) [0x0000d] in <64887e5183394bbb92d4e93b00abc715>:0 
  at Mono.Cecil.ModuleDefinition.ReadModule (System.String fileName, Mono.Cecil.ReaderParameters parameters) [0x00083] in <64887e5183394bbb92d4e93b00abc715>:0 
  at Mono.Cecil.AssemblyDefinition.ReadAssembly (System.String fileName, Mono.Cecil.ReaderParameters parameters) [0x00000] in <64887e5183394bbb92d4e93b00abc715>:0 
  at Program.RewriteAssembly (System.String assemblyLocation, System.Collections.Generic.Dictionary`2[TKey,TValue] resourcesStrings, Program+CmdOptions options) [0x00102] in <e22a64c33f534be5973f5a15969f1a9e>:0 
  at Program.Main (System.String[] args) [0x00108] in <e22a64c33f534be5973f5a15969f1a9e>:0 
[ERROR] FATAL UNHANDLED EXCEPTION: Mono.Cecil.Cil.SymbolsNotMatchingException: Symbols were found but are not matching the assembly
  at Mono.Cecil.ModuleDefinition.ReadSymbols (Mono.Cecil.Cil.ISymbolReader reader, System.Boolean throwIfSymbolsAreNotMaching) [0x00045] in <64887e5183394bbb92d4e93b00abc715>:0 
  at Mono.Cecil.ModuleReader.ReadSymbols (Mono.Cecil.ModuleDefinition module, Mono.Cecil.ReaderParameters parameters) [0x00064] in <64887e5183394bbb92d4e93b00abc715>:0 
  at Mono.Cecil.ModuleReader.CreateModule (Mono.Cecil.PE.Image image, Mono.Cecil.ReaderParameters parameters) [0x0008e] in <64887e5183394bbb92d4e93b00abc715>:0 
  
```

but then retry doesn't do that.

--- a/mcs/mcs/cs-parser.jay
+++ b/mcs/mcs/cs-parser.jay
@@ -1586,7 +1586,7 @@ constructor_body
 	 expression SEMICOLON
 	 {
 		lexer.parsing_block = 0;
-		current_block.AddStatement (new ContextualReturn ((Expression) $3));
+		current_block.AddStatement (CreateExpressionBodiedStatement ((Expression) $3));
 		var b = end_block (GetLocation ($4));
 		b.IsCompilerGenerated = true;
 		$$ = b;
@@ -1607,7 +1607,7 @@ expression_block
 	 lambda_arrow_expression SEMICOLON
 	 {
 		lexer.parsing_block = 0;
-		current_block.AddStatement (new ContextualReturn ((Expression) $3));
+		current_block.AddStatement (CreateExpressionBodiedStatement ((Expression) $3));
 		var b = end_block (GetLocation ($4));
 		b.IsCompilerGenerated = true;
 		$$ = b;
@@ -8028,6 +8028,14 @@ static bool IsUnaryOperator (Operator.OpType op)
 		return true;
 	}
 	return false;
+}
+
+static Statement CreateExpressionBodiedStatement (Expression expr)
+{
+	if (expr is ThrowExpression te)
+		return new StatementExpression (te);
+
+	return new ContextualReturn (expr);
 }
 
 void syntax_error (Location l, string msg)

--- a/mcs/mcs/statement.cs
+++ b/mcs/mcs/statement.cs
@@ -928,8 +928,7 @@ namespace Mono.CSharp {
 		public override Reachability MarkReachable (Reachability rc)
 		{
 			base.MarkReachable (rc);
-			expr.MarkReachable (rc);
-			return rc;
+			return expr.MarkReachable (rc);
 		}
 
 		public override bool Resolve (BlockContext ec)

--- a/mcs/tests/ver-il-net_4_x.xml
+++ b/mcs/tests/ver-il-net_4_x.xml
@@ -73699,7 +73699,7 @@
         <size>32</size>
       </method>
       <method name="Int32 Test()" attrs="134">
-        <size>10</size>
+        <size>2</size>
       </method>
       <method name="System.Object Foo()" attrs="129">
         <size>10</size>
@@ -73708,16 +73708,16 @@
         <size>23</size>
       </method>
       <method name="Void Test3(Int32 ByRef)" attrs="145">
-        <size>3</size>
+        <size>2</size>
       </method>
       <method name="Int32 get_Item(Int32)" attrs="2177">
-        <size>10</size>
+        <size>2</size>
       </method>
       <method name="Void add_Event(System.Action)" attrs="2182">
-        <size>3</size>
+        <size>2</size>
       </method>
       <method name="Void remove_Event(System.Action)" attrs="2182">
-        <size>3</size>
+        <size>2</size>
       </method>
       <method name="Void TestExpr_1(Boolean)" attrs="129">
         <size>21</size>
@@ -73744,7 +73744,7 @@
         <size>7</size>
       </method>
       <method name="Int32 TestExpr_6(Int32 ByRef)" attrs="145">
-        <size>10</size>
+        <size>2</size>
       </method>
       <method name="Int32 TestExpr_7(Int32 ByRef)" attrs="129">
         <size>15</size>

